src/compiler/checker.ts

```diff
@@ -8823,26 +8869,144 @@ module ts {
             }
 
-            var exprType = checkExpression(node.expression);
+            var rightType = checkExpression(node.expression);
             // unknownType is returned i.e. if node.expression is identifier whose name cannot be resolved
             // in this case error about missing name is already reported - do not report extra one
-            if (!allConstituentTypesHaveKind(exprType, TypeFlags.Any | TypeFlags.ObjectType | TypeFlags.TypeParameter)) {
+            if (!allConstituentTypesHaveKind(rightType, TypeFlags.Any | TypeFlags.ObjectType | TypeFlags.TypeParameter)) {
                 error(node.expression, Diagnostics.The_right_hand_side_of_a_for_in_statement_must_be_of_type_any_an_object_type_or_a_type_parameter);
             }
 
             checkSourceElement(node.statement);
         }
 
+        function checkForInOrForOfVariableDeclaration(iterationStatement: ForInStatement | ForOfStatement): void {
+            var variableDeclarationList = <VariableDeclarationList>iterationStatement.initializer;
+            if (variableDeclarationList.declarations.length >= 1) {
+                var decl = variableDeclarationList.declarations[0];
+                checkVariableDeclaration(decl);
+            }
+        }
+
+        function getTypeForVariableDeclarationInForOfStatement(forOfStatement: ForOfStatement): Type {
+            // Temporarily return 'any' below ES6
+            if (languageVersion < ScriptTarget.ES6) {
+                return anyType;
+            }
+
+            // iteratedType will be undefined if the for-of expression type was missing properties/signatures
+            // required to get it's iteratedType (like [Symbol.iterator] or next). This may be
+            // because we accessed properties from anyType, or it may have led to an error inside
+            // getIteratedType.
+            return getIteratedType(getTypeOfExpression(forOfStatement.expression), forOfStatement.expression) || anyType;
+        }
+
+        /**
+         * When expressionForError is undefined, it means we should not report any errors.
+         */
+        function getIteratedType(iterable: Type, expressionForError: Expression): Type {
+            Debug.assert(languageVersion >= ScriptTarget.ES6);
+            var iteratedType = getIteratedTypeSubroutine(iterable, expressionForError);
+            // Now even though we have extracted the iteratedType, we will have to validate that the type
+            // passed in is actually an Iterable.
+            if (expressionForError && iteratedType) {
+                var completeIterableType = globalIterableType !== emptyObjectType ? createTypeReference(<GenericType>globalIterableType, [iteratedType]) : emptyObjectType;
+                checkTypeAssignableTo(iterable, completeIterableType, expressionForError);
+            }
+
+            return iteratedType;
+            
+            function getIteratedTypeSubroutine(iterable: Type, expressionForError: Expression) {

```

Comment:
It's not really a helper, it's actually doing all the heavy lifting, so it's the main worker. I can call it getIteratedType_worker, or the more ambiguous getIteratedTypeWorker for those allergic to underscores. But it really is here just for the sake of getIteratedType, and as long as it's a worker, no one else should see it. The only reason I use a subroutine (as opposed to a block) is so that `return` both sets the result and leaves the block early.